feat(cli): Add setting to show status(or Gemini 's thoughts) in terminal title and taskbar icon

Fridayxiao opened the pull request:
## TLDR

This feature provides users with ambient feedback on the CLI's current operational status by dynamically updating the terminal window's title. This is especially useful when the window is minimized, as the status is visible in the taskbar.



## Dive Deeper

Adds a new setting showStatusInTitle to display the CLI's current operation (e.g., "Analyzing files...") in the terminal title bar. This provides users with at-a-glance status updates,  especially when the window is in the background. When the CLI is idle, the original terminal  title is restored.

### Changes
 - packages/cli/src/config/settingsSchema.ts: Added the showStatusInTitle: boolean setting to the UI
  category
  - packages/cli/src/config/settings.ts: Added setting mapping for configuration access
  - packages/cli/src/ui/AppContainer.tsx:
    - Added useEffect hook to dynamically update terminal title based on streamingState, current thought,
  and the showStatusInTitle setting
    - Integrates with existing computeWindowTitle utility function
    - Respects both showStatusInTitle and hideWindowTitle settings
    - Handles various operational states (Idle, Active, WaitingForConfirmation)
  - packages/cli/src/ui/AppContainer.test.tsx: Added comprehensive test coverage for all title update
  scenarios
  - packages/cli/src/utils/windowTitle.ts: Utility function for computing window titles with proper
  sanitization

- docs/cli/configuration.md: Documented the new showStatusInTitle setting with a description
  and example.

## Reviewer Test Plan

1. Enable the feature:
    - Add "showStatusInTitle": true to your ~/.gemini/settings.json file under the ui section
    - Start the CLI (npm start). The terminal title should show the default Gemini title
  2. Verify status updates:
    - Run a command that takes a few seconds (e.g., ask to analyze a large codebase)
    - Observe that the terminal title changes to reflect the current status and thoughts
    - Look for titles like "Thinking: Analyzing project structure..."
  3. Verify idle state restoration:
    - Once the command completes, confirm that the terminal title reverts back to the original title
  4. Verify integration with hideWindowTitle:
    - Set "hideWindowTitle": true in settings
    - Verify that terminal title doesn't change even with showStatusInTitle: true
  5. Disable the feature:
    - Set "showStatusInTitle": false in your settings.json
    - Restart the CLI and run another command
    - Confirm that the terminal title does not change during the operation
  6. Check the documentation:
    - Review the new entry for showStatusInTitle in the settings schema to ensure it's clear and accurate

## Testing Matrix


|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ❓  |  ✅  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

Related to #3437

Resolves #4373 

## Rendering
<img width="170" height="41" alt="image" src="https://github.com/user-attachments/assets/76b6461f-168c-49a4-b97f-39962b59c548" />
<img width="178" height="43" alt="image" src="https://github.com/user-attachments/assets/cf30e412-a4d7-46b1-9326-a49819b33abf" />

jacob314 commented:
Sorry we missed this PR. This is an interesting change that I would approve if the following changes are made
Show the status based on a settings specified in settings.ts rather than a command line arg. That is our preferred way of surfacing optional settings like this that users would tend to always want enabled if they happen to like the setting.

Fridayxiao commented:
> Sorry we missed this PR. This is an interesting change that I would approve if the following changes are made Show the status based on a settings specified in settings.ts rather than a command line arg. That is our preferred way of surfacing optional settings like this that users would tend to always want enabled if they happen to like the setting.

I had passed the preflight checks, but I could not pass the E2E tests, because my Gemini 's API quota is not enough

Fridayxiao commented:
/gemini reivew

Fridayxiao commented:
/gemini review

Fridayxiao commented on packages/cli/src/ui/App.tsx:
```diff
@@ -730,6 +730,46 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
+    }
+
+    // Pad the title to a fixed width to prevent taskbar icon resizing.
+    const paddedTitle = title.padEnd(80, ' ');
```
It is impossible for the taskbar icon to be longer than 80 characters, so this would not cause the icon to resize

Fridayxiao commented:
Hi, @jacob314  I'm looking forward to my PR being accepted and contributing to the project. Could you please review the code when you have time? Thanks

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -730,6 +730,46 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
+
+  // Effect to update the terminal title with the current status.
+  useEffect(() => {
+    if (!settings.merged.showStatusInTitle) return;
```
See setWindowTitle in gemini.tsx which has similar logic. It would be good to combine the two so we attempt a single operation to cleanup the title at the end of running Gemini CLI. Also note that you should respect hideWindowTitle as that method does.

Fridayxiao commented:
Thank you for review! @jacob314  The code now :
1. uses a single useEffect hook to manage the title, respects the existing hideWindowTitle setting,
2. removed the extra cleanup useEffect to maintain a single operation to cleanup the title at the end of running Gemini CLI 
3. removed extra text "Working..." text and loading phrases
4. The ANSI escape code was changed from "\x1b]0;" to "\x1b]2;", match the setWindowTitle function
5. added a unittest

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -783,6 +783,35 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
+
+  // Effect to update the terminal title with the current status.
+  useEffect(() => {
+    // Respect both showStatusInTitle and hideWindowTitle settings (like setWindowTitle)
```
remove this comment.

jacob314 commented on packages/cli/src/ui/App.test.tsx:
```diff
@@ -1675,4 +1688,300 @@ describe('App UI', () => {
+        />,
+      );
+
+      // Check that no title-related writes occurred (only bracketed paste mode writes)
```
comment about bracketed paste mode is not needed. remove

jacob314 reviewed (commented):
Thanks for the fixes. This is approved after the following comments are updated. Not that settingsSchema has been updated so take care merging to reflect those changes.

Fridayxiao commented:
Hi @jacob314 , could you please review my changes? I am still looking forward to make some contribution to the gemini CLI, Thank you.

jacob314 commented:
Please sync to the latest and then I will review.
Right now when I try to run locally I get the following errors.
```

src/ui/App.tsx:850:26 - error TS2339: Property 'showStatusInTitle' does not exist on type 'InferSettings<{ readonly mcpServers: { readonly type: "object"; readonly label: "MCP Servers"; readonly category: "Advanced"; readonly requiresRestart: true; readonly default: Record<string, MCPServerConfig>; readonly description: "Configuration for MCP servers."; readonly showInDialog: false; }; ... 13 more ...; re...'.

850     if (!settings.merged.showStatusInTitle || settings.merged.hideWindowTitle)
                             ~~~~~~~~~~~~~~~~~

src/ui/App.tsx:850:63 - error TS2339: Property 'hideWindowTitle' does not exist on type 'InferSettings<{ readonly mcpServers: { readonly type: "object"; readonly label: "MCP Servers"; readonly category: "Advanced"; readonly requiresRestart: true; readonly default: Record<string, MCPServerConfig>; readonly description: "Configuration for MCP servers."; readonly showInDialog: false; }; ... 13 more ...; re...'.

850     if (!settings.merged.showStatusInTitle || settings.merged.hideWindowTitle)
                                                                  ~~~~~~~~~~~~~~~

src/ui/App.tsx:873:21 - error TS2339: Property 'showStatusInTitle' does not exist on type 'InferSettings<{ readonly mcpServers: { readonly type: "object"; readonly label: "MCP Servers"; readonly category: "Advanced"; readonly requiresRestart: true; readonly default: Record<string, MCPServerConfig>; readonly description: "Configuration for MCP servers."; readonly showInDialog: false; }; ... 13 more ...; re...'.

873     settings.merged.showStatusInTitle,
                        ~~~~~~~~~~~~~~~~~

src/ui/App.tsx:874:21 - error TS2339: Property 'hideWindowTitle' does not exist on type 'InferSettings<{ readonly mcpServers: { readonly type: "object"; readonly label: "MCP Servers"; readonly category: "Advanced"; readonly requiresRestart: true; readonly default: Record<string, MCPServerConfig>; readonly description: "Configuration for MCP servers."; readonly showInDialog: false; }; ... 13 more ...; re...'.

874     settings.merged.hideWindowTitle,
                        ~~~~~~~~~~~~~~~
                        ```

Fridayxiao commented:
Great!, Thanks a lot! @jacob314

Fridayxiao commented:
Hi @jacob314 , I just synced to the latest and resolved a conflict, this seems to have caused the auto-merge to be turned off.  Could you please tell me what I should do?  Thank you! 😀
I would also like to know when it will be merged. Do I need to go through all the workflows like E2E Tests?  I had passed the preflight checks, but I could not pass the E2E tests, because my Gemini 's API quota is not enough

jacob314 commented:
Got it back in the merge queue.

Fridayxiao commented:
Sorry, I forgot to run the preflight checks after I resolved the conflict, let me fix this

Fridayxiao commented:
Hey, I have ran the preflight and tested the feature manually, it worked, I think it's ready. Waitting for your consent. My friends and I are looking forward to this new feature. Please review when you have time, @jacob314 , Thanks.

Fridayxiao commented on packages/cli/src/ui/AppContainer.tsx:
```diff
@@ -976,6 +982,40 @@ Logging in with Google... Please restart Gemini CLI to continue.
+    if (streamingState === StreamingState.Idle) {
+      title = originalTitleRef.current;
+    } else {
+      const statusText = thought?.subject;
```
Gemini's suggestion looks good.

jacob314 reviewed (approved):
![lgtm](https://i.imgur.com/JgNRalI.png)
